test/testactivitylistmodel.cpp

```diff
@@ -0,0 +1,416 @@
+{
+    Q_OBJECT
+
+public:
+    TestActivityListModel() = default;
+    ~TestActivityListModel() override {
+        OCC::AccountManager::instance()->deleteAccount(accountState.data());
+    }
+
+    QScopedPointer<FakeQNAM> fakeQnam;
+    OCC::AccountPtr account;
+    QScopedPointer<OCC::AccountState> accountState;
+
+    OCC::Activity testNotificationActivity;
+
+    static constexpr int searchResultsReplyDelay = 100;
+
+private slots:
+    void initTestCase()
+    {
+        fakeQnam.reset(new FakeQNAM({}));
+        account = OCC::Account::create();
+        account->setCredentials(new FakeCredentials{fakeQnam.data()});
+        account->setUrl(QUrl(("http://example.de")));
+
+        accountState.reset(new OCC::AccountState(account));
+
+        fakeQnam->setOverride([this](QNetworkAccessManager::Operation op, const QNetworkRequest &req, QIODevice *device) {
+            Q_UNUSED(device);
+            QNetworkReply *reply = nullptr;
+
+            const auto urlQuery = QUrlQuery(req.url());
+            const auto format = urlQuery.queryItemValue(QStringLiteral("format"));
+            const auto since = urlQuery.queryItemValue(QStringLiteral("since")).toInt();
+            const auto limit = urlQuery.queryItemValue(QStringLiteral("limit")).toInt();
+            const auto path = req.url().path();
+
+            if (!req.url().toString().startsWith(accountState->account()->url().toString())) {
+                reply = new FakeErrorReply(op, req, this, 404, fake404Response);
+            }
+            if (format != QStringLiteral("json")) {
+                reply = new FakeErrorReply(op, req, this, 400, fake400Response);
+            }
+
+            if (path.startsWith(QStringLiteral("/ocs/v2.php/apps/activity/api/v2/activity"))) {
+                reply = new FakePayloadReply(op, req, FakeRemoteActivityStorage::instance()->activityJsonData(since, limit), searchResultsReplyDelay, fakeQnam.data());
+            }
+
+            if (!reply) {
+                return qobject_cast<QNetworkReply*>(new FakeErrorReply(op, req, this, 404, QByteArrayLiteral("{error: \"Not found!\"}")));
+            }
+
+            return reply;
+        });
+
+        OCC::AccountManager::instance()->addAccount(account);
+
+        // Activity comparison is done by checking type, id, and accName
+        // We need an activity with these details, at least
+        testNotificationActivity._accName = accountState->account()->displayName();
+        testNotificationActivity._id = 1;
+        testNotificationActivity._type = OCC::Activity::NotificationType;
+        testNotificationActivity._dateTime = QDateTime::currentDateTime();
+    };
+
+    // Test receiving activity from server
+    void testFetchingRemoteActivity() {
+        TestingALM model;
+        model.setAccountState(accountState.data());
+        QAbstractItemModelTester modelTester(&model);
+
+        QCOMPARE(model.rowCount(), 0);
+
+        model.startFetchJob();
```

Comment:
so here there is a problem because you call a protected method (OK in practice you override it with a public one) and this makes the test use a different API and interface than real code
this is not going to work well and can lead to behavior differences between real code and tests

Reply:
Ingesting activities into the activity list model requires either doing it via a `startFetchJob` which calls the server, which we override here, or directly through `activitiesReceived` which takes in a JSON document as an argument and parses it -- but this method is also protected, so we'd need to create a wrapper to access this too. 

The overriden `startFetchJob` slightly increases test coverage as it is also testing the signal that invoked `activitiesReceived`, hence using that. We can't use the default  `startFetchJob` as it checks the `accountState` to see if it is connected